### ISSUE TITLE
Binary release installation script fails randomly in Github Actions

### ISSUE THREAD

**Issue by douglasiacovelli** [firebase/firebase-tools#1874]
### [REQUIRED] Environment info

**firebase-tools:** latest

**Platform:** Ubuntu latest on Github Actions

### [REQUIRED] Test case

To run `curl -sL firebase.tools | bash` on github actions

### [REQUIRED] Steps to reproduce

Create a workflow on github actions and execute `curl -sL firebase.tools | bash`

### [REQUIRED] Expected behavior

To install firebase tools on Github Actions. The first time it stopped working was around `2019-12-17 23:50:00 UTC`

### [REQUIRED] Actual behavior

It hangs without completing the install. Later the CI halts by timeout

```
2019-12-19T13:44:46.4242037Z ##[group]Run curl -vL firebase.tools | bash
2019-12-19T13:44:46.4243179Z [36;1mcurl -vL firebase.tools | bash[0m
2019-12-19T13:44:46.4279145Z shell: /bin/bash -e {0}
2019-12-19T13:44:46.4280018Z env:
2019-12-19T13:44:46.4281442Z   JAVA_HOME: /opt/hostedtoolcache/jdk/8.0.232/x64
2019-12-19T13:44:46.4282462Z   JAVA_HOME_8.0.232_x64: /opt/hostedtoolcache/jdk/8.0.232/x64
2019-12-19T13:44:46.4283496Z ##[endgroup]
2019-12-19T13:44:46.4443199Z * Rebuilt URL to: firebase.tools/
2019-12-19T13:44:46.4444348Z   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
2019-12-19T13:44:46.4445607Z                                  Dload  Upload   Total   Spent    Left  Speed
2019-12-19T13:44:46.4446599Z 
2019-12-19T13:44:46.5692898Z   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0*   Trying 151.101.1.195...
2019-12-19T13:44:46.5694200Z * TCP_NODELAY set
2019-12-19T13:44:46.5743572Z * Connected to firebase.tools (151.101.1.195) port 80 (#0)
2019-12-19T13:44:46.5745037Z > GET / HTTP/1.1
2019-12-19T13:44:46.5745900Z > Host: firebase.tools
2019-12-19T13:44:46.5747128Z > User-Agent: curl/7.58.0
2019-12-19T13:44:46.5747975Z > Accept: */*
2019-12-19T13:44:46.5748810Z > 
2019-12-19T13:44:46.5797200Z < HTTP/1.1 301 Moved Permanently
2019-12-19T13:44:46.5798123Z < Server: Varnish
2019-12-19T13:44:46.5801960Z < Retry-After: 0
2019-12-19T13:44:46.5803150Z < Location: https://firebase.tools/
2019-12-19T13:44:46.5804395Z < Content-Length: 0
2019-12-19T13:44:46.5805798Z < Accept-Ranges: bytes
2019-12-19T13:44:46.5806882Z < Date: Thu, 19 Dec 2019 13:44:46 GMT
2019-12-19T13:44:46.5807777Z < Connection: close
2019-12-19T13:44:46.5808844Z < X-Served-By: cache-bwi5076-BWI
2019-12-19T13:44:46.5810086Z < X-Cache: HIT
2019-12-19T13:44:46.5811217Z < X-Cache-Hits: 0
2019-12-19T13:44:46.5812292Z < X-Timer: S1576763087.573176,VS0,VE0
2019-12-19T13:44:46.5813185Z < 
2019-12-19T13:44:46.5813619Z 
2019-12-19T13:44:46.5815227Z   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
2019-12-19T13:44:46.5816367Z * Closing connection 0
2019-12-19T13:44:46.5817490Z * Issue another request to this URL: 'https://firebase.tools/'
2019-12-19T13:44:46.5838331Z *   Trying 151.101.65.195...
2019-12-19T13:44:46.5839334Z * TCP_NODELAY set
2019-12-19T13:44:46.5889492Z * Connected to firebase.tools (151.101.65.195) port 443 (#1)
2019-12-19T13:44:46.5897509Z * ALPN, offering h2
2019-12-19T13:44:46.5898386Z * ALPN, offering http/1.1
2019-12-19T13:44:46.5969645Z * successfully set certificate verify locations:
2019-12-19T13:44:46.5971396Z *   CAfile: /etc/ssl/certs/ca-certificates.crt
2019-12-19T13:44:46.5972349Z   CApath: /etc/ssl/certs
2019-12-19T13:44:46.5973190Z } [5 bytes data]
2019-12-19T13:44:46.5974037Z * TLSv1.3 (OUT), TLS handshake, Client hello (1):
2019-12-19T13:44:46.5975043Z } [512 bytes data]
2019-12-19T13:44:46.6052824Z * TLSv1.3 (IN), TLS handshake, Server hello (2):
2019-12-19T13:44:46.6053768Z { [106 bytes data]
2019-12-19T13:44:46.6054663Z * TLSv1.2 (IN), TLS handshake, Certificate (11):
2019-12-19T13:44:46.6055581Z { [4434 bytes data]
2019-12-19T13:44:46.6056906Z * TLSv1.2 (IN), TLS handshake, Server key exchange (12):
2019-12-19T13:44:46.6059510Z { [300 bytes data]
2019-12-19T13:44:46.6062986Z * TLSv1.2 (IN), TLS handshake, Server finished (14):
2019-12-19T13:44:46.6063984Z { [4 bytes data]
2019-12-19T13:44:46.6067460Z * TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
2019-12-19T13:44:46.6069655Z } [37 bytes data]
2019-12-19T13:44:46.6073041Z * TLSv1.2 (OUT), TLS change cipher, Client hello (1):
2019-12-19T13:44:46.6075378Z } [1 bytes data]
2019-12-19T13:44:46.6078052Z * TLSv1.2 (OUT), TLS handshake, Finished (20):
2019-12-19T13:44:46.6078998Z } [16 bytes data]
2019-12-19T13:44:46.6134598Z * TLSv1.2 (IN), TLS handshake, Finished (20):
2019-12-19T13:44:46.6135569Z { [16 bytes data]
2019-12-19T13:44:46.6136963Z * SSL connection using TLSv1.2 / ECDHE-RSA-AES128-GCM-SHA256
2019-12-19T13:44:46.6139520Z * ALPN, server accepted to use h2
2019-12-19T13:44:46.6140778Z * Server certificate:
2019-12-19T13:44:46.6142007Z *  subject: CN=roomq.io
2019-12-19T13:44:46.6142867Z *  start date: Nov 19 22:53:01 2019 GMT
2019-12-19T13:44:46.6143773Z *  expire date: Feb 17 22:53:01 2020 GMT
2019-12-19T13:44:46.6147348Z *  subjectAltName: host "firebase.tools" matched cert's "firebase.tools"
2019-12-19T13:44:46.6148680Z *  issuer: C=US; O=Let's Encrypt; CN=Let's Encrypt Authority X3
2019-12-19T13:44:46.6149708Z *  SSL certificate verify ok.
2019-12-19T13:44:46.6152902Z * Using HTTP2, server supports multi-use
2019-12-19T13:44:46.6153863Z * Connection state changed (HTTP/2 confirmed)
2019-12-19T13:44:46.6154890Z * Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
2019-12-19T13:44:46.6157448Z } [5 bytes data]
2019-12-19T13:44:46.6160845Z * Using Stream ID: 1 (easy handle 0x55a6f75716b0)
2019-12-19T13:44:46.6163323Z } [5 bytes data]
2019-12-19T13:44:46.6165511Z > GET / HTTP/2
2019-12-19T13:44:46.6166439Z > Host: firebase.tools
2019-12-19T13:44:46.6167641Z > User-Agent: curl/7.58.0
2019-12-19T13:44:46.6168493Z > Accept: */*
2019-12-19T13:44:46.6169329Z > 
2019-12-19T13:44:46.6208322Z { [5 bytes data]
2019-12-19T13:44:46.6209314Z * Connection state changed (MAX_CONCURRENT_STREAMS updated)!
2019-12-19T13:44:46.6210692Z } [5 bytes data]
2019-12-19T13:44:46.7845149Z 
2019-12-19T13:44:46.8621135Z   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0< HTTP/2 200 
2019-12-19T13:44:46.8622507Z < cache-control: max-age=3600
2019-12-19T13:44:46.8623579Z < content-type: text/html; charset=utf-8
2019-12-19T13:44:46.8624562Z < etag: ce13fe09b0b53f816cceea31c29d0ab9e8008b61817e5aec3cebab40ea5b04f4
2019-12-19T13:44:46.8626223Z < last-modified: Tue, 17 Dec 2019 17:45:00 GMT
2019-12-19T13:44:46.8628451Z < strict-transport-security: max-age=31556926
2019-12-19T13:44:46.8629618Z < accept-ranges: bytes
2019-12-19T13:44:46.8630969Z < date: Thu, 19 Dec 2019 13:44:46 GMT
2019-12-19T13:44:46.8632475Z < x-served-by: cache-bwi5038-BWI
2019-12-19T13:44:46.8633536Z < x-cache: MISS
2019-12-19T13:44:46.8634656Z < x-cache-hits: 0
2019-12-19T13:44:46.8635708Z < x-timer: S1576763087.619564,VS0,VE240
2019-12-19T13:44:46.8636879Z < vary: x-fh-requested-host, accept-encoding
2019-12-19T13:44:46.8637903Z < content-length: 7169
2019-12-19T13:44:46.8638867Z < 
2019-12-19T13:44:46.8639775Z { [2480 bytes data]
2019-12-19T13:44:46.8642572Z 
2019-12-19T13:44:46.8644008Z 100  7169  100  7169    0     0  17150      0 --:--:-- --:--:-- --:--:-- 91910
2019-12-19T13:44:46.8645002Z * Connection #1 to host firebase.tools left intact
```

**Comment by abeisgoat** [comment]
Did it work for awhile then stop working?

**Comment by douglasiacovelli** [comment]
It used to work all the time and then it stopped working even though we did nothing. We've found a workaround which was to run the inner commands of the script manually: `sudo curl -o "/usr/local/bin/firebase" -L https://firebase.tools/bin/linux/latest && sudo chmod +rx "/usr/local/bin/firebase"`

**Comment by abeisgoat** [comment]
So there is a known issue with the way `firebase.tools` works where immediately post-release we have about a 10 minute outage where all calls to the script will fail because the latest release on Github doesn't have binaries attached to it yet. I think this is what you were hitting and since your work-around, by your own admission, is to do the same thing the script does by hand, that would imply that you just wasted a little time then it worked (because the builds existed). It's a bit hard to tell from the curl log, but the fact you're getting a `x-cache: MISS` implies that the binary isn't there yet, so really I just need to improve how to flag when a release is ready to the installer script.

**Comment by douglasiacovelli** [comment]
That seems a bit odd, @abeisgoat because we certainly had an outage of more than 5~10 minutes. It stopped working for days. It only began to work again after the workaround. It's hardly a coincidence in my opinion because I had two simultaneous open PRs: one running with the script as described on the website and another one with the workaround and the first kept giving me the same results.

**Comment by abeisgoat** [comment]
Hmm okay, thanks for the extra info, I'll do some Github Actions specific investigation as this is quite strange.

**Comment by alex-benoit** [comment]
I can confirm this happened as well for me
on `unbuntu-latest` image, `curl -sL https://firebase.tools | bash` hangs for 1h+ with no action :(

**Comment by SudoPlz** [comment]
Any workarounds until then?

**Comment by alex-benoit** [comment]
you can download using npm
`run: sudo npm install -g firebase-tools`

**Comment by abeisgoat** [comment]
Hey folks, sorry for the delay, frankly I was just hoping this would fix itself as Github Actions evolved. There's some issue with the fact that `firebase.tools` downloads the binaries from GIthub and Github Actions blocks that in some way. For us to fix this we'd basically need to host the binaries somewhere else, which is pretty outside our current flow. This is pretty stupid and terrible though, so I'd still like to fix it, but it's a bit of a project.

**Comment by samtstern** [comment]
@abeisgoat can/should we report this to GitHub somehow?

**Comment by abeisgoat** [comment]
Fixed in #3429
